content/post/2020-02-04-code-generation/index.md

```diff
@@ -0,0 +1,237 @@
+_What_ happens exactly when the package is installed from source? 
+A configure script is run ([configure](https://github.com/civisanalytics/civis-r/blob/master/configure) or [configure.win](https://github.com/civisanalytics/civis-r/blob/master/configure.win)). 
+Such scripts are automatically run when building a package from source. 
+Here's what this script does: sourcing `tools/run_generate_client.R`.
+
+```r
+"${R_HOME}"/bin/Rscript tools/run_generate_client.R
+```
+
+And the script called sources R code from the package that fetches the API spec and writes code and `roxygen2` docs in `R/generated_client.R`. 
+When the package is _not_ installed from source, the users get the `R/generated_client.R` that's last been generated by the package maintainer, so if the Civis platform itself was updated in the meantime, the users might find a platform endpoint is missing from the `civis` package.
+The approach used by `civis` has the clear advantage of allowing a perfect synchronization between the wrapped platform and the package.
+
+#### Creating functions lists and R6 methods in `minicss`
+
+In [`mimicss`](https://github.com/coolbutuseless/minicss/) by mikefc, ["Lists of CSS property information is turned into function lists and R6 methods."](https://twitter.com/coolbutuseless/status/1199458806092034049). 
+See [aaa.R](https://github.com/coolbutuseless/minicss/blob/fe378f6e040405e51fb07cb74fd2f0bac8a85b26/R/aaa.R) and [prop_transform.R](https://github.com/coolbutuseless/minicss/blob/fe378f6e040405e51fb07cb74fd2f0bac8a85b26/R/prop_transform.R). 
+As in most examples the code is generated as a string, but in that case it's not written to disk, it becomes code via the use of [`eval()` and `parse()`](http://adv-r.had.co.nz/Expressions.html#parsing-and-deparsing).
+
+#### Generate C++ bindings with `Rcpp::compileAttributes()`
+
+`Rcpp::compileAttributes()` generates code (the bindings required to call C++ functions from R) after scanning a package source files. Find more information [in the `Rcpp` vignette about attributes](https://cran.r-project.org/web/packages/Rcpp/vignettes/Rcpp-attributes.pdf). You could call the function "whenever functions areadded, removed, or have their signatures changed." but the aforementioned vignette also states " if you are using either RStudio or `devtools`to build your package then the `compileAttributes` function is called automatically whenever your package is built".
+
+`Rcpp::compileAttributes()` is with `scaffolder` one of the only "general" solutions mentioned in this post, as opposed as other approaches developed for and within a given package.
+
+### Generating code on-the-fly
+
+One step further, one might generate code on-the-fly, i.e. as users run the package.
+
+* The [`chromote`](https://github.com/rstudio/chromote) package ["generates auto-completable R6 methods at runtime"](https://twitter.com/alandipert/status/1199456564442062848).
```

Comment:
What does "on-the-fly" mean? When you load the package? When you run a function? When you create an R6 class? When you create an R6 object?